docs/markdown/Fs-module.md

```diff
@@ -194,3 +194,32 @@ Returns the last component of the path, dropping the last part of the suffix
 fs.stem('foo/bar/baz.dll')  # baz
 fs.stem('foo/bar/baz.dll.a')  # baz.dll
 ```
+
+### relative_to
+
+*since 0.57.0*
+
+Given two absolute paths, returns a version of the first path relative to the second.
+If a path is given as the 'within' argument, the function will return the first path
+unchanged if it is not inside 'within'.
+
+The `relative_to` function is system dependent. It accepts the `native: true` argument
+to compute relative paths on the build system. Otherwise it uses the host system.
+
+
+Examples: 
+
+```meson
+# On windows:
+fs.relative_to('c:\\proj1\\foo', 'c:\\proj1\\bar')  # '..\\foo'
+fs.relative_to('c:\\proj1\\foo', 'd:\\proj1\\bar')  # ERROR, since the relative path does not exist
+fs.relative_to('c:\\proj1\\foo', 'd:\\proj1\\bar', within: 'd:\\')  # 'c:\\proj1\\foo', within allows to get an absolute path
```

Comment:
This functionality is good, but the UI as it stands is a bit confusing. I had to read the documentation several times in order to understand what it does and why, Now this might just be a question of me being stupid, but it would seem prudent to try to get a nicer experience.

As an example, if the kwarg were called `if_within` or `is_within` its meaning would be more clear. It would indicate a condition rather than a general parameter of some sort.

Secondly, this is problematic in the Windows case if you don't directly control the paths yourself. If you get the file strings as arguments from somewhere, with values `c:\foo` and g:\bar` you can't say that an absolute path is fine without extracting the drive letter as a substring. This is unnecessary work for users. Having a kwarg like `allow_absolute_if_drive_letters_are_different` would be simpler.

Finally, the `native` approach seems a bit limited as people may need to build different kinds of paths so it might be better to instead have a kwarg `path_type` with the values `build`, `host`, `unix` and `windows` with `host` being the default.

Reply:
`fs` seems to be more obviously related to native paths. It may make sense to have path-related methods on the various machine objects, but that would be a different change (instead, I don't see the benefit in hard-coding that one is dealing with POSIX or Windows paths).

Regarding the syntax, I'd propose something like

```
fs.relative_to('c:\\proj1\\foo', 'c:\\proj1\\bar')  # '..\\foo'
fs.relative_to('c:\\proj1\\foo', 'c:\\proj2\\bar', within: 'c:\\proj2)  # ERROR, the relative path is outside the "within" argument
fs.relative_to('c:\\proj1\\foo', 'd:\\proj1\\bar')  # ERROR, for Windows the default "within" is the drive
fs.relative_to('c:\\proj1\\foo', 'c:\\proj2\\bar', within: 'c:\\proj2', allow_absolute: true)  # 'c:\\proj1\\foo'
fs.relative_to('c:\\proj1\\foo', 'd:\\proj1\\bar', allow_absolute: true)  # 'c:\\proj1\\foo'
```

Reply:
@bonzini Do we need the `within` if we have `allow_absolute`?

Reply:
Ah yes, I see what you expect to have with each of those arguments

Reply:
In my opinion, both the `within` and `allow_absolute` arguments are a consequence of other design decisions of the meson language:

- `allow_absolute` would not be necessary if it was easy to do something like:

```
try:
    something = fs.relative_to(path1, path2)
except:
    something = path1
```

- `within` would not be necessary if it was easy to do:

```
if fs.is_relative_to(path1, within):
    path1 = path1.relative_to(path2)
```

Are you sure you want to use arguments in `fs.relative_to()` to implement those features instead of working towards a `try/except` and an `is_relative_to()` functionality? It seems to me you are trying to fix a language issue with arguments, but maybe I miss some overarching goal

Reply:
This is the `relative_to` current behaviour from a functional perspective. I have implemented most of your suggestions:

### relative_to

*since 0.64.0*

Given two paths, returns a version of the first path relative to the second.

Examples: 

```meson
# On windows:
fs.relative_to('c:\\proj1\\foo', 'c:\\proj1\\bar')  # '..\\foo'
fs.relative_to('c:\\proj1\\foo', 'd:\\proj1\\bar')  # ERROR, since the relative path does not exist

# On other systems:
fs.relative_to('/prefix/lib', '/prefix/bin')  # '../lib'
fs.relative_to('/prefix/lib/foo', '/prefix')  # 'lib/foo'
```

`relative_to` has some optional keyword arguments:

- `native`: Can be set to `true` or `false`. Whether the relative paths are computed
  following the rules of the build machine (native: `true`) or the host (native: `false`).

- `allow_absolute`: If set to `true`, instead of giving an error, `relative_to()`
  will return the original path.  Useful if an absolute path is a reasonable
  fallback.

- `if_within`: A path. `relative_to(path1, path2, within: path3)` returns
  the relative path of `path1 with respect to `path2` if `path1` is found in `path3`,
  otherwise it returns `path1` unchanged.

Examples: 

```meson
# On windows:
fs.relative_to('c:\\proj1\\foo', 'c:\\proj2\\bar', if_within: 'c:\\proj2')  # ERROR, the relative path is outside the "within" argument
fs.relative_to('c:\\proj1\\foo', 'c:\\proj2\\bar', if_within: 'c:\\proj2', allow_absolute: true)  # 'c:\\proj1\\foo'
fs.relative_to('c:\\proj1\\foo', 'd:\\proj1\\bar', allow_absolute: true)  # 'c:\\proj1\\foo'

# On other systems:
fs.relative_to('/project1/lib/foo', '/usr/bin', if_within: '/usr')  # ERROR: '/project1/lib/foo' not in '/usr'
fs.relative_to('/project1/lib/foo', '/usr/bin', if_within: '/usr', allow_absolute: 'true')  # '/project1/lib/foo'
fs.relative_to('/usr/lib/foo', '/usr/bin', if_within: '/usr')  # '../lib/foo'
```